### PR TITLE
Feat/327 add image upload and phone pic

### LINKED ISSUE #330: Add upload feature to AddFoodItem
Add the functionality such that the user is able to add their own pictures to represent the foodItem if they are discontent with the choices given by the FoodFacts API. Additionally store them in Firebase Storage.

### PR DESCRIPTION
**Pull request by frostymate**
This pull request adds to the functionality of #266 in which you are now able to also upload pictures from your phone's storage so that images that cannot be fetched from the FoodFacts API can be tailored to the User's request. I had to set up a storage bucket and add the necessary functions in order to write to the firebase storage in the FoodItem Model. Additionally I found out that to upload an image to the Firebase storage it took a few seconds so I added some circular indicator and make the user unable to submit while he is uploading an image. Additionally I had originally planned to also have it such that the user could also take pictures from the phone camera and I got it to work but the app crashed if you did not give it the right permisions. Which we have done beforehand for the scanner, however I think that I would need to modularize and adapt the CameraPersmisionHandler and as we are nearing the end of the project, I don't have enough times and there are more things that are in priority. Anyway something for the future that I would also liked to add was that right now, if you delete a foodItem, if the image was uploaded from the phone, it is not deleted from the firebase storage which is a problem so that would be something to work on in the future.

### Dependencies and Permissions:

* [`app/build.gradle.kts`](diffhunk://#diff-8cff73265af19c059547b76aca8882cbaa3209291406f52df1dafbbc78e80c46R295): Added `firebase.storage` dependency.
* [`app/src/main/AndroidManifest.xml`](diffhunk://#diff-7fa6aef292187a049f7a4d6060d8df3ba212d838789c78940bd363344b1c38cdR11-R12): Added permissions for reading and writing external storage.

### Repository Enhancements:

* [`app/src/main/java/com/android/shelfLife/model/foodItem/FoodItemRepository.kt`](diffhunk://#diff-1d39032c8d5dcb32ca37bb4c947228c1796b7a9c7ab9c231f3e7e9e974a4083bR53-R56): Added methods `setIsGenerated` and `uploadImageToFirebaseStorage`.
* [`app/src/main/java/com/android/shelfLife/model/foodItem/FoodItemRepositoryFirestore.kt`](diffhunk://#diff-134c98d56f4f23ac654e2c4128a182cf4f2186b51a980b63aa31f673ef961afbR233-R252): Implemented `setIsGenerated` and `uploadImageToFirebaseStorage` methods.

### CODE REVIEW

**Comment by frostymate** [comment]
I would like to add that this PR does not pass the CI because we wanted this feature pushed for BOTA. Tests will be done in a seperate task

File: app/src/main/java/com/android/shelfLife/model/foodItem/FoodItemRepository.kt [hunk 1]
```diff
@@ -47,4 +50,8 @@ interface FoodItemRepository {
   fun selectFoodItem(foodItem: FoodItem?)
 
   fun setFoodItems(householdId: String, value: List<FoodItem>)
+
+  fun setIsGenerated(value: Boolean)
```
Comment (matvitn) [inline app/src/main/java/com/android/shelfLife/model/foodItem/FoodItemRepository.kt]: This should not be called IsGenerated. We are not generating a fooditem, we are quick adding it

File: app/src/main/java/com/android/shelfLife/model/foodItem/FoodItemRepositoryFirestore.kt [hunk 2]
```diff
@@ -25,6 +31,8 @@ class FoodItemRepositoryFirestore @Inject constructor(private val db: FirebaseFi
   private val _errorMessage = MutableStateFlow<String?>(null)
   override val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()
 
+  private val _isGenerated = MutableStateFlow<Boolean>(false)
+  override val isGenerated: StateFlow<Boolean> = _isGenerated.asStateFlow()
```
Comment (matvitn) [inline app/src/main/java/com/android/shelfLife/model/foodItem/FoodItemRepositoryFirestore.kt]: Name change here too

File: app/src/main/java/com/android/shelfLife/ui/overview/AddFoodItem.kt [hunk 3]
```diff
@@ -146,88 +140,6 @@ fun AddFoodItemScreen(
                 Spacer(modifier = Modifier.height(32.dp))
               }
 
-              //              if (foodFacts.isNotEmpty()) {
```
Comment (matvitn) [inline app/src/main/java/com/android/shelfLife/ui/overview/AddFoodItem.kt:149]: Thanks for the clean-up!!!

File: app/src/main/java/com/android/shelfLife/ui/overview/ChooseFoodItem.kt [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+      }
+
+    LaunchedEffect(foodItemViewModel.capturedImageUri) {
+        foodItemViewModel.capturedImageUri?.let { uri ->
+            foodItemViewModel.isLoading = true
+            val uploadedUrl = foodItemViewModel.uploadImageToFirebaseStorage(uri, context)
+            if (uploadedUrl != null) {
+                foodItemViewModel.selectedImage = FoodFacts(
+                    name = "",
+                    quantity = Quantity(1.0, FoodUnit.GRAM),
```
Comment (matvitn) [inline app/src/main/java/com/android/shelfLife/ui/overview/ChooseFoodItem.kt:77]: It'd be good to add some comments here to specify that this is a temporary food item that is used for the next screen

File: app/src/main/java/com/android/shelfLife/ui/overview/ChooseFoodItem.kt [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+                    Text(
+                        text = stringResource(id = R.string.image_not_loading_error),
+                        color = MaterialTheme.colorScheme.error,
+                        modifier = Modifier.padding(16.dp))
+                  }
+                  else -> Unit
+                }
+              }
+
+              item { Spacer(modifier = Modifier.height(8.dp)) }
```
Comment (matvitn) [inline app/src/main/java/com/android/shelfLife/ui/overview/ChooseFoodItem.kt:232]: why did you put a spacer in an item box 😂😂😂😂😂

File: app/src/main/java/com/android/shelfLife/ui/overview/FirstFoodItem.kt [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    foodItemViewModel: FoodItemViewModel = hiltViewModel(),
+) {
+
+  val coroutineScope = rememberCoroutineScope()
+  val context = LocalContext.current
+
+  Scaffold(
+      topBar = {
+        CustomTopAppBar(
+            onClick = { navigationActions.goBack() },
```
Comment (matvitn) [inline app/src/main/java/com/android/shelfLife/ui/overview/FirstFoodItem.kt]: You need to reset the selectedfooditem (set it to null) and call foodItemViewModel.resetSearchStatus()

File: app/src/main/java/com/android/shelfLife/ui/overview/FirstFoodItem.kt [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+              item {
+                FoodNameField(
+                    foodItemViewModel.foodName,
+                    onFoodNameChange = { newValue -> foodItemViewModel.changeFoodName(newValue) },
+                    foodNameErrorResId = foodItemViewModel.foodNameErrorResId,
+                    modifier = Modifier.padding(bottom = 16.dp).testTag("inputFoodName"))
+              }
+              item {
+                CustomButtons(
+                    button1OnClick = {
```
Comment (matvitn) [inline app/src/main/java/com/android/shelfLife/ui/overview/FirstFoodItem.kt:71]: same issue here as for the go back button

File: app/src/main/java/com/android/shelfLife/ui/utils/ExtendedFloatingButtons.kt [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+import androidx.compose.ui.graphics.vector.ImageVector
+import androidx.compose.ui.platform.testTag
+import androidx.compose.ui.unit.dp
+import androidx.hilt.navigation.compose.hiltViewModel
+import com.android.shelfLife.ui.navigation.NavigationActions
+import com.android.shelfLife.ui.navigation.Screen
+import com.android.shelfLife.viewmodel.overview.FoodItemViewModel
+
+@Composable
+fun ExtendedActionButtons(
```
Comment (matvitn) [inline app/src/main/java/com/android/shelfLife/ui/utils/ExtendedFloatingButtons.kt:27]: Thank you for refactoring!!! this reduces a lot of duplication and allows future use as well!!! great job.

File: app/src/main/java/com/android/shelfLife/ui/utils/UIComponents.kt [hunk 9]
```diff
@@ -200,3 +207,47 @@ fun CustomButtons(
             }
       }
 }
+
+@Composable
+fun ExpandableFAB(
```
Comment (matvitn) [inline app/src/main/java/com/android/shelfLife/ui/utils/UIComponents.kt]: This is unnecessary due to the other extended FAB composable you created

File: app/src/main/java/com/android/shelfLife/viewmodel/overview/FoodItemViewModel.kt [hunk 10]
```diff
@@ -207,6 +248,38 @@ constructor(
...
+
+  fun resetSelectFoodItem() {
+    foodItemRepository.selectFoodItem(
+        FoodItem(
+            uid = "",
+            foodFacts = FoodFacts(name = "", quantity = Quantity(0.0, FoodUnit.GRAM)),
+            owner = ""))
+  }
+
+  fun setIsGenerated(isGenerated: Boolean) {
```
Comment (matvitn) [inline app/src/main/java/com/android/shelfLife/viewmodel/overview/FoodItemViewModel.kt]: Change the name, same issue as for the others (shouldnt be isGenerated)

File: app/src/main/java/com/android/shelfLife/viewmodel/overview/IndividualFoodItemViewModel.kt [hunk 11]
```diff
@@ -23,6 +23,14 @@ constructor(
     selectedFood = foodItemRepository.selectedFoodItem.value
   }
 
+  fun unselectFoodItem() {
+    foodItemRepository.selectFoodItem(null)
+  }
+
+  fun setIsGenerated(isGenerated: Boolean) {
```
Comment (matvitn) [inline app/src/main/java/com/android/shelfLife/viewmodel/overview/IndividualFoodItemViewModel.kt]: Same issue here, shouldnt be isGenerated

**Review by matvitn — CHANGES_REQUESTED** [review]
After 2 weeks we finally got this feature on the App!!!!!! its been a long (extremelly long) process but im happy that this is almost ready for main.

The overall code is very clean!! There are some conceptual choices that I would've done differently but your choice is also valid (and most importantly it works!!!).

I also appreciate you changing the tests so that they pass with your version! We really need that coverage right now so i appreciate the effort.

I also really appreciate the refactoring you did for the FAB to reduce code duplication and increase modularity!


Unfortunately I will have to extend the duration of this process by a little. There are a few things I would like you to change:
-Change all the isGenerated names to something else (something like isQuickAdd). We arent generating a fooditem, we are simply doing a quick add with an image.
-Add some inline comments to explain when you set a temporary fooditem (in choose fooditem and first fooditem for example)
-There are specific parts (there are comments for it) where you havent reset selected fooditem and search status properly leading to bugs (like going to the scanner and having a previously non-added banana show up as a scanned item). This viewmodel/model is very intricate and requires careful handling to prevent bugs
-Delete the previous (unused) extended FAB class to reduce lines to cover and ensure unused composables get deleted from the project


Overall this PR is so close to being done (I can almost taste it!!) it just needs a few adjustments before its ready for main

Let me know as soon as you make the changes (or if youd like me to clarify something for you) and ill review this PR again (and hopefuly merge it into main)!!!

**Review by matvitn — APPROVED** [review]
Thanks for the changes!!!!! This PR is ready to merge.

Approved!

**Pull request merged**